File: src/ast/unit.zig
const SourceLoc = @import("../parser/token.zig").SourceLoc;
pub const MapDecl = @import("map.zig").MapDecl; // re-export map type

pub const Unit = struct {
    name: []const u8,
    loc: SourceLoc,
    sections: []const []const u8,
    license: ?[]const u8,
    body: []Stmt,
};

pub const Stmt = struct {
    kind: StmtKind,
    loc: SourceLoc,
};

pub const StmtKind = union(enum) {
    Return: *Expr,
    VarDecl: VarDecl,
    HeapVarDecl: HeapVarDecl,
    Assignment: Assignment,
    IfGuard: IfGuard,
};

pub const Assignment = struct {
    target: *Expr,
    op: []const u8,
    value: *Expr,
};

pub const IfGuard = struct {
    condition: Expr,
    body: []Stmt,
};

pub const Expr = struct {
    kind: ExprKind,
    loc: SourceLoc,
};

pub const ExprKind = union(enum) {
    VarRef: []const u8,
    Number: i64,
    MethodCall: MethodCall,
    HeapLookup: HeapLookup,
    Dereference: *Expr,
};

pub const VarDecl = struct {
    name: []const u8,
    var_type: VarType,
    value: *Expr,
};

pub const VarType = enum {
    reg,
    imm,
};

pub const HeapLookup = struct {
    map_name: []const u8,
    key_expr: *Expr,
};

pub const MethodCall = struct {
    receiver: []const u8,
    method: []const u8,
    arg: *Expr,
};

pub const HeapVarDecl = struct {
    name: []const u8,
    lookup: HeapLookup,
};

File: src/ast/unit.rs
use crate::parser::SourceLoc;

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct Unit {
    pub name: String,
    pub loc: SourceLoc,
    pub sections: Vec<String>,
    pub license: Option<String>,
    pub body: Vec<Stmt>,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct Stmt {
    pub kind: StmtKind,
    pub loc: SourceLoc,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub enum StmtKind {
    Return(Box<Expr>),
    VarDecl(VarDecl),
    HeapVarDecl(HeapVarDecl),
    Assignment(Assignment),
    IfGuard(IfGuard),
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct Assignment {
    pub target: Box<Expr>,
    pub op: AssignmentOp,
    pub value: Box<Expr>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[allow(unused)]
pub enum AssignmentOp {
    Assign,     
    AddAssign,  
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct IfGuard {
    pub condition: Expr,
    pub body: Vec<Stmt>,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct Expr {
    pub kind: ExprKind,
    pub loc: SourceLoc,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub enum ExprKind {
    Variable(String),
    Number(i64),
    MethodCall(MethodCall),
    HeapLookup(HeapLookup),
    Dereference(Box<Expr>),
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct VarDecl {
    pub name: String,
    pub var_type: VarType,
    pub value: Box<Expr>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[allow(unused)]
pub enum VarType {
    Reg,  
    Imm,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct HeapLookup {
    pub map_name: String,
    pub key_expr: Box<Expr>,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct MethodCall {
    pub receiver: String,
    pub method: String,
    pub arg: Box<Expr>,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct HeapVarDecl {
    pub name: String,
    pub lookup: HeapLookup,
}

